ResearchKit/ActiveTasks/ORKStroopStepViewController.m

```diff
@@ -253,31 +257,62 @@ - (void)startNextQuestionOrFinish {
 }
 
 - (void)startQuestion {
+    BOOL isText;
+    UIColor *textColor = nil;
+    ORKStroopStep *stroopStep = (ORKStroopStep *)self.step;
+    switch (stroopStep.stroopStyle) {
+        case ORKStroopStyleText:
+            isText = YES;
+            break;
+        case ORKStroopStyleTextAlwaysBlack:
+            isText = YES;
+            textColor = [UIColor blackColor];
+            break;
+        case ORKStroopStyleBox:
+            isText = NO;
+            break;
+        case ORKStroopStyleRandomizeTextOrBox: {
+            int textOrBox = arc4random() % 2;
+            isText = textOrBox;
+            if (isText) {
+                textColor = [UIColor blackColor];
+            }
+        }
+        default:
+            break;
+    }
+
     if ([self stroopStep].randomizeVisualAndColorAlignment) {
         int pattern = arc4random() % 2;
         if (pattern == 0) {
+            // matching color for text and color
             int index = arc4random() % [self.colors.allKeys count];
             NSString *text = [self.colors.allKeys objectAtIndex:index];
             self.stroopContentView.colorLabelText = text;
-            UIColor *color = [self.colors valueForKey:text];
-            self.stroopContentView.colorLabelColor = color;
+            UIColor *color = textColor ?: [self.colors valueForKey:text];
+            [self.stroopContentView setColor:color isText:isText];
+            _correctColorString = text;
         }
         else {
+            // non-matching color for text and color
             int index = arc4random() % [self.differentColorLabels.allKeys count];
             NSString *text = [self.differentColorLabels.allKeys objectAtIndex:index];
             self.stroopContentView.colorLabelText = text;
             NSArray *colorArray = [self.differentColorLabels valueForKey:text];
             int randomColor = arc4random() % colorArray.count;
             UIColor *color = [colorArray objectAtIndex:randomColor];
-            self.stroopContentView.colorLabelColor = color;
+            [self.stroopContentView setColor:color isText:isText];
+            _correctColorString = [self.colors allKeysForObject:color][0];
         }
     } else {
         int index = arc4random() % [self.colors.allKeys count];
         NSString *text = [self.colors.allKeys objectAtIndex:index];
+        _correctColorString = text;
         self.stroopContentView.colorLabelText = text;
-        UIColor *color = [self.colors valueForKey:text];
-        self.stroopContentView.colorLabelColor = color;
+        UIColor *color = textColor ?: [self.colors valueForKey:text];
+        [self.stroopContentView setColor:color isText:isText];
     }
+    _singleTestStroopStyle = isText ? @"text" : @"box";
     [self setButtonsEnabled];
     _startTime = [NSProcessInfo processInfo].systemUptime;
```

Comment:
This is the logic table for StroopStyle and RandomizeVisualAndColorAlignment. Is this correct? ORKStroopStyleTextAlwaysBlack will sometimes show non-black text. Also this seems very complex.

StroopStyle | Randomize | Behavior
-|-|-
ORKStroopStyleBox|False| random RGBY box
ORKStroopStyleBox|True| random RGBY box
ORKStroopStyleText|False| random RGBY color with matching RGBY text
ORKStroopStyleText|True| random RGBY color with random RGBY text
ORKStroopStyleTextAlwaysBlack|False| black color with random RGBY text
ORKStroopStyleTextAlwaysBlack|True| (black color with random RGBY text) or (random RGBY color with non-matching random RGBY text)
ORKStroopStyleRandomizeTextOrBox|False| (random RGBY box) or (black color with random RGBY text)
ORKStroopStyleTextAlwaysBlack|True| (random RGBY box) or (black color with random RGBY text) or (random RGBY color with non-matching random RGBY text)

Reply:
I was initially trying to keep the code minimally different than what came from the stable branch. But I agree that this is hard to read now with the multiple options. Have refactored into a clearer flow now.

Reply:
I like the changes. LGTM